src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts

```diff
@@ -0,0 +1,139 @@
+ *--------------------------------------------------------------------------------------------*/
+
+import { CancellationToken } from '../../../../../base/common/cancellation.js';
+import { localize } from '../../../../../nls.js';
+import { IWorkspaceContextService } from '../../../../../platform/workspace/common/workspace.js';
+import { IExplorerService } from '../../../files/browser/files.js';
+import { CountTokensCallback, IToolData, IToolImpl, IToolInvocation, IToolResult } from '../languageModelToolsService.js';
+import { ExplorerItem } from '../../../files/common/explorerModel.js';
+import { SortOrder } from '../../../files/common/files.js';
+import { IToolInputProcessor } from './tools.js';
+
+/**
+ * Represents either a file (string) or a directory (tuple with string and children).
+ */
+export type DirectoryItem = string | [string, DirectoryItem[]];
+
+export const ExtensionProjectTreeId = 'positron_getProjectTree';
+export const InternalProjectTreeId = `${ExtensionProjectTreeId}_internal`;
+export const ProjectTreeToolData: IToolData = {
+	id: InternalProjectTreeId,
+	displayName: localize('chat.tools.getProjectTree', "Get Project Tree"),
+	modelDescription: `Construct a tree of the directories and files in the project.`,
+	tags: ['positron-assistant']
+};
+
+export class ProjectTreeTool implements IToolImpl {
+	constructor(
+		@IWorkspaceContextService private readonly _workspaceContextService: IWorkspaceContextService,
+		@IExplorerService private readonly _explorerService: IExplorerService,
+	) { }
+
+	async invoke(invocation: IToolInvocation, countTokens: CountTokensCallback, token: CancellationToken): Promise<IToolResult> {
+		const workspaceFolders = this._workspaceContextService.getWorkspace().folders;
+		if (workspaceFolders.length === 0) {
+			return {
+				content: [],
+				toolResultMessage: 'No workspace folders found.'
+			};
+		}
+
+		const workspaceTrees: DirectoryItem[][] = [];
+		const treeErrors: string[] = [];
+		const sortOrder = this._explorerService.sortOrderConfiguration.sortOrder;
+
+		for (const workspaceFolder of workspaceFolders) {
+			const explorerRoot = this._explorerService.findClosestRoot(workspaceFolder.uri);
+			if (!explorerRoot) {
+				treeErrors.push(`No explorer root found for ${workspaceFolder.name}`);
+				continue;
+			}
+			try {
+				const tree = await constructDirectoryTree(explorerRoot, sortOrder);
+				if (!tree) {
+					continue;
+				}
+				workspaceTrees.push(tree);
+			} catch (error) {
+				treeErrors.push(`Failed to generate tree for ${workspaceFolder.name}: ${error}`);
+			}
+		}
+
+		return {
+			content: workspaceTrees.map(dirTree => ({ kind: 'text', value: JSON.stringify(dirTree) })),
+			toolResultMessage: treeErrors.length > 0 ? treeErrors.join('\n') : undefined
+		};
+	}
+}
+
+export class ProjectTreeInputProcessor implements IToolInputProcessor {
+	processInput(input: any) {
+		// No input processing needed for this tool
+		return input;
+	}
+}
+
+const IGNORED_DIRECTORIES = [
```

Comment:
This can be a future improvement: I think there is existing functionality in the editor for ignoring files in certain contexts, which may also use `.gitignore`. Would we want to hook into that?

Reply:
Yes, I think it would be helpful to hook into that! Maybe an optional input parameter to indicate whether the project tree should be constructed with or without gitignored files? I added a TODO note inline. There's also the `files.exclude` setting that we likely want to read as well.

I could see a user being confused about missing files and the "full" project tree could help reveal which files are hidden. At the same time, probably for most use cases, we don't need to include the ignored files.